backend/de.metas.handlingunits.base/src/main/java/org/adempiere/mm/attributes/spi/impl/HUVendorBPartnerAttributeValuesProvider.java

```diff
@@ -0,0 +1,302 @@
+
+	}
+
+	@Override
+	public boolean isHighVolume()
+	{
+		return true;
+	}
+
+	@Override
+	public boolean isAllowAnyValue()
+	{
+		return true;
+	}
+
+	@Override
+	public Evaluatee prepareContext(final IAttributeSet attributeSet)
+	{
+		final I_M_HU hu = huAttributesBL.getM_HU(attributeSet);
+		final int bpartnerId = hu.getC_BPartner_ID();
+		final int huId = hu.getM_HU_ID();
+
+		return Evaluatees.mapBuilder()
+				.put(CTXNAME_M_HU_ID, huId > 0 ? huId : -1)
+				.put(CTXNAME_CurrentVendor_BPartner_ID, bpartnerId > 0 ? bpartnerId : -1)
+				.build();
+	}
+
+	@Override
+	public List<? extends NamePair> getAvailableValues(final Evaluatee evalCtx)
+	{
+		final int huId = CTXNAME_M_HU_ID.getValueAsInteger(evalCtx);
+		final int currentVendorId = CTXNAME_CurrentVendor_BPartner_ID.getValueAsInteger(evalCtx);
+
+		//
+		// Create the cache key based on M_HU_ID and current vendor
+		final ArrayKey cacheKey = ArrayKey.of(
+				huId <= 0 ? System.currentTimeMillis() : huId,      // if M_HU_ID <= 0 then make it unique
+				currentVendorId <= 0 ? -1 : currentVendorId // normalize
+		);
+
+		//
+		// Get from cache / Load
+		return hu2Vendors.getOrLoad(cacheKey, () -> retrieveVendorKeyNamePairs(currentVendorId));
+	}
+
+	@Nullable
+	private static String normalizeValueKey(final Object valueKey)
+	{
+		if (valueKey == null)
+		{
+			return null;
+		}
+		else if (valueKey instanceof Number)
+		{
+			final int valueKeyAsInt = ((Number)valueKey).intValue();
+			return String.valueOf(valueKeyAsInt);
+		}
+		else
+		{
+			return valueKey.toString();
+		}
+	}
+
+	@Override
+	@Nullable
+	public NamePair getAttributeValueOrNull(final Evaluatee evalCtx, final Object valueKey)
+	{
+		final List<? extends NamePair> availableValues = getAvailableValues(evalCtx);
+		if (availableValues.isEmpty())
+		{
+			return null;
+		}
+
+		// Normalize the valueKey
+		final String valueKeyNormalized = normalizeValueKey(valueKey);
+
+		for (final NamePair vnp : availableValues)
+		{
+			if (Objects.equals(vnp.getID(), valueKeyNormalized))
+			{
+				return vnp;
+			}
+		}
+
+		return null;
+	}
+
+	@Nullable
+	@Override
+	public AttributeValueId getAttributeValueIdOrNull(final Object valueKey)
+	{
+		return null;
+	}
+
+	/**
+	 * Retrieve allowed vendor KeyNamePairs.
+	 *
+	 * Mainly it will contain following values:
+	 * <ul>
+	 * <li>{@link #getNullValue()}
+	 * <li>available vendors
+	 * <li>current HU's vendor if not present in the list above
+	 * </ul>
+	 *
+	 * @param currentVendorBPartnerId current vendor ID
+	 * @return available values
+	 */
+	private static List<KeyNamePair> retrieveVendorKeyNamePairs(
+			final int currentVendorBPartnerId)
+	{
+		final Properties ctx = Env.getCtx();
+
+		final List<KeyNamePair> vendors = new ArrayList<>();
+		vendors.add(staticNullValue());
+
+		final List<KeyNamePair> vendorsCached = HUVendorBPartnerAttributeValuesProvider.vendors.getOrLoad(currentVendorBPartnerId, () -> retrieveVendorKeyNamePairs());
```

Comment:
this is fishy?!

why caching the same `retrieveVendorKeyNamePairs()` for each `currentVendorBPartnerId`?